#1 Importing Libraries

In [32]:
# Import Liabraries
import pandas as pd
import numpy as np
import os

In [33]:
# define path
path=r'/Users/yeji/Documents/Career Foundry/Achievement 4/Instacart Basket Analysis'

In [34]:
# define orders_wrangled csv
df_ords= pd.read_csv(os.path.join(path,'02 Data','Prepared Data','orders_wrangled.csv'), index_col = 0)

In [35]:
# define producs csv
df_prods = pd.read_csv(os.path.join(path,'02 Data', 'Original Data', 'products.csv'), index_col = 0)

In [36]:
# create a dataframe
df_test=pd.DataFrame()

In [37]:
# create a mixed type column
df_test['mix']=['a','b',1, True]

In [38]:
df_test

mix
0     a
1     b
2     1
3  True

In [39]:
# check for mix type
for col in df_test.columns.tolist(): weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
if len (df_test[weird]) > 0: print (col)


mix


In [40]:
# change data type to string
df_test['mix'] = df_test['mix'].astype('str')

In [41]:
# find missing value in df_prods
df_prods.isnull().sum()

product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [42]:
# create dataframe with missing values
df_nan = df_prods[df_prods['product_name'].isnull()==True]

In [43]:
df_nan

product_name  aisle_id  department_id  prices
product_id                                              
34                  NaN       121             14    12.2
69                  NaN        26              7    11.8
116                 NaN        93              3    10.8
262                 NaN       110             13    12.1
525                 NaN       109             11     1.2
1511                NaN        84             16    14.3
1780                NaN       126             11    12.3
2240                NaN        52              1    14.2
2586                NaN       104             13    12.4
3159                NaN       126             11    13.1
3230                NaN       120             16    14.4
3736                NaN        41              8    14.8
4283                NaN        77              7    14.4
4790                NaN        91             16    14.5
38183               NaN        39             12    20.9
40440               NaN       120             16    14.8

In [44]:
df_prods.shape

(49693, 4)

In [45]:
# create new data frame without missing values for df_prods
df_prods_clean = df_prods[df_prods['product_name'].isnull()==False]

In [46]:
df_prods_clean.shape

(49677, 4)

In [47]:
# new data frame with duplicates only in df_prods
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [48]:
df_dups

product_name  aisle_id  \
product_id                                                                
462                         Fiber 4g Gummy Dietary Supplement        70   
18458                                              Ranger IPA        27   
26808                    Black House Coffee Roasty Stout Beer        27   
35306       Gluten Free Organic Peanut Butter & Chocolate ...       121   
35491                                 Adore Forever Body Wash       127   

            department_id  prices  
product_id                         
462                    11     4.8  
18458                   5     9.2  
26808                   5    13.4  
35306                  14     6.8  
35491                  11     9.9

In [49]:
# dropping duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [50]:
df_prods_clean_no_dups.shape

(49672, 4)

#2 Data consistency check on dr_ords

In [51]:
#2. running describe in df_ords
df_ords.describe()

order_id       user_id  order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02        6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

the describe function shows that in order_hour_of_day, the min is 0 while maxis 23 but the 50% range is 13, which is lower than the actual 50% of 23 which is 11.5. Also, the max of days_since_prior_order is 30 but 50% is 7, which is also lower than the actual 50% of 16. 

In [52]:
#3 checking for mixed data type in df_ords
for col in df_ords.columns.tolist():weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
if len (df_ords[weird]) > 0: print (col)

#4 no mixed data type was found in the columns

In [53]:
#5 checking for missing values
df_ords.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

#5 days_since_prior_order has 206209 missing values. this is probably these users had no prior ordering before and it was their first orders

In [54]:
# create new data frame with only missing values for df_ords
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull()==True]

In [55]:
df_ords_nan

order_id  user_id  order_number  orders_day_of_week  \
0         2539329        1             1                   2   
11        2168274        2             1                   2   
26        1374495        3             1                   1   
39        3343014        4             1                   6   
45        2717275        5             1                   3   
...           ...      ...           ...                 ...   
3420930    969311   206205             1                   4   
3420934   3189322   206206             1                   3   
3421002   2166133   206207             1                   6   
3421019   2227043   206208             1                   1   
3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 6 columns]

In [56]:
df_ords_nan.shape

(206209, 6)

In [57]:
#6 imputing '0' for missing values in days_since_prior_order
df_ords['days_since_prior_order'].fillna(0, inplace=True)

In [58]:
#checking if missing values were turned to 0
df_ords.isnull().sum()

order_id                  0
user_id                   0
order_number              0
orders_day_of_week        0
order_hour_of_day         0
days_since_prior_order    0
dtype: int64

I changed the missing values in days_since_prior_order to 0 because the 0 indicates that those users never had a prior order before. 

In [59]:
#check for duplicates
df_ords_dups = df_ords[df_ords.duplicated()]

In [60]:
df_ords_dups

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

no duplicates were found 

In [61]:
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))

In [62]:
df_prods_clean_no_dups.to_csv(os.path.join(path,'02 Data','Prepared Data','products_checked.csv'))